distsys/resources/nestedarch.go

```diff
@@ -0,0 +1,344 @@
+		if tpe.Equal(expectedTpe) {
+			return nil
+		}
+	}
+	return fmt.Errorf("%w: archetype response had unexpected tag %v, expected %v", ErrNestedArchetypeProtocol, tpe, expectedTpes)
+}
+
+func (res *nestedArchetype) Abort() chan struct{} {
+	res.assertSanity(false)
+	go func() {
+		staleAcksReceived := 0
+
+		retryReq:
+		select {
+		case res.sendCh <-tla.MakeTLARecord([]tla.TLARecordField{
+			{nestedArchetypeConstants.tpe, nestedArchetypeAbortReq },
+		}):
+			// go to next select, we successfully sent the abort request
+		case resp := <-res.receiveCh:
+			err := res.handleResponseValue(resp, false, nestedArchetypeReadAck, nestedArchetypeWriteAck)
+			if err != nil {
+				panic(err)
+			}
+			if staleAcksReceived != 0 {
+				panic(fmt.Errorf("received a second stale read/write ack while trying to perform abort: %v", resp))
+			}
+			staleAcksReceived++
+			goto retryReq
+		case err := <-res.ctxErrCh:
+			panic(fmt.Errorf("encountered error while trying to abort critical section: %w", err))
+		}
+
+		select {
+		case resp := <-res.receiveCh:
+			err := res.handleResponseValue(resp, false, nestedArchetypeAbortAck)
+			if err != nil {
+				panic(err)
+			}
+		case err := <-res.ctxErrCh:
+			panic(fmt.Errorf("encountered error while trying to abort critical section: %w", err))
+		}
+
+		res.cachedStructCh <- struct{}{}
+	}()
+	return res.cachedStructCh
+}
+
+func (res *nestedArchetype) PreCommit() chan error {
+	res.assertSanity(true)
+	go func() {
+		resp, err := res.performRequestOrAbort(nestedArchetypePreCommitReq)
+		if err != nil {
+			res.cachedErrCh <- err
+			return
+		}
+		res.cachedErrCh <- res.handleResponseValue(resp, true, nestedArchetypePreCommitAck)
+	}()
+	return res.cachedErrCh
+}
+
+func (res *nestedArchetype) Commit() chan struct{} {
+	res.assertSanity(true)
+	go func() {
+		resp, err := res.performRequest(nestedArchetypeCommitReq)
+		if err != nil {
+			panic(err)
+		}
+		err = res.handleResponseValue(resp, false, nestedArchetypeCommitAck)
+		if err != nil {
+			panic(err)
+		}
+		res.cachedStructCh <- struct{}{}
+	}()
+	return res.cachedStructCh
+}
+
+func (res *nestedArchetype) ReadValue() (_ tla.TLAValue, err error) {
+	res.assertSanity(true)
+	defer res.catchTLATypeErrors(&err)
+
+	resp, err := res.performRequestOrAbort(nestedArchetypeReadReq)
+	if err != nil {
+		return tla.TLAValue{}, err
+	}
+	err = res.handleResponseValue(resp, true, nestedArchetypeReadAck)
+	if err != nil {
+		return tla.TLAValue{}, err
+	}
+	return resp.ApplyFunction(nestedArchetypeConstants.value), nil
+}
+
+func (res *nestedArchetype) WriteValue(value tla.TLAValue) (err error) {
+	res.assertSanity(true)
+	defer res.catchTLATypeErrors(&err)
+
+	resp, err := res.performRequestOrAbort(nestedArchetypeWriteReq,
+		tla.TLARecordField{Key: nestedArchetypeConstants.value, Value: value})
+	if err != nil {
+		return err
+	}
+	return res.handleResponseValue(resp, true, nestedArchetypeWriteAck)
+}
+
+func (res *nestedArchetype) Close() (err error) {
```

Comment:
I see your point that in this case you don't want to distinguish the execution and cleanup errors. However, still I think it's not the general case. So now you're just aggregating the error that makes sense in this case. Also, with the old API, you just have to write one more for loop here to aggregate cleanup errors separately. I know that there is a code duplication here, but the API simplicity and consistency worth it.

Reply:
I'm still not convinced either... maybe a simple way to describe what I see as the problem: imagine that, when you called a function in Go, execution of the `defer` statements inside was optional, and you had to remember to write `call.Close()` if execution of `defer` statements was important to you. I basically see `ctx.Run()` (or the variant I'm proposing, which is its own conversation) as analogous to an FFI function call to MPCal, and... well, having to "close" a function call doesn't make sense to me, and doesn't seem simpler.

I remembered your comparison between `distsys.MPCalContext` and the standard Go `context.Context` (which is what I think this is about), and I think they're... just not the same. `context.Context` provides a way to do sophisticated things with something like a "request", and I can see how it's useful for writing server code in that way. `distsys.MPCalContext`, the way I see it, is basically like an FFI to an MPCal archetype's execution. As a user, you configure the call, and you either launch and wait for termination, or run it in a daemon-like mode (expect run forever, maybe kill it later). In neither case do I really care about what this "FFI call" actually does internally (even resource cleanup; in normal Go, resources that last only during the function call should be cleaned up internally, and we mash all the errors into a single `error` all the time), other than: most simply, "did it work correctly?"; or in a more complex situation "stop now please".

To help users confuse these two "contexts" directly is actually bad imo, because they work very differently. After writing this, I am now convinced that your suggestion while I was working to change to a different name than `Close` was a very important idea.

If we really wanted `context.Context` support (assuming that archetype lifetime was related to some request lifetime), then we'd probably want an adapter like this, which (see comments) might require situation-specific details:

```
func RunWithContext(ctx context.Context, dctx *distsys.MPCalContext) error {
    errCh := make(chan error)
    go func() {
        // debatable whether we want Run or RunDiscardingExits here...
        // cancellation error will come from elsewhere anyway, depends
        // if we care whether we exit normally vs were killed
        errCh <- dctx.Run()
    }()
    select {
    case <-ctx.Done():
        dctx.RequestExit()
        return multierr.Append(<-errCh, ctx.Err())
    case err := <-errCh:
        return err // may be nil; this is the successful exit path
    }
}
```

If we wrote this same adapter (which we would still need) with the existing API, the code would just be more complicated (at least the ability to forget `Close`, more below) with no apparent benefit:
```
func RunWithContext(ctx context.Context, dctx *distsys.MPCalContext) error {
    errCh := make(chan error)
    go func() {
        // debatable whether we want Run or RunDiscardingExits here...
        // cancellation error will come from elsewhere anyway, depends
        // if we care whether we exit normally vs were killed
        errCh <- dctx.Run()
    }()
    select {
    case <-ctx.Done():
        err := dctx.Close()
        return multierr.Combine(<-errCh, err, ctx.Err())
    case err := <-errCh:
        err = multierr.Append(err, dctx.Close())
        return err // may be nil; this is the successful exit path
    }
}
```
When writing this, I not only had to write more code, but I was also confused by where to write `dctx.Close()`. My instincts said "put in defer", but `defer` usually goes next to `dctx := NewMPCalContext(...)`, which is not here.
However, putting `Close` in `defer` or outside this API adapter (in cleanup code area?) puts part of the cancellation semantics in a really strange place: the first `select` branch would be `return nil`, probably with a "go look for the `defer dctx.Close()`, it's here somewhere..."; in this simple example I just gave two reasonable places someone might try to put it. You could also try to encapsulate the `NewMPCalContext(...)` part, but that requires a bunch of parameter forwarding, and the `Close` semantics are still strange, because it's really "kill process, and oh by the way do all the clean-up, so you have to call this even if you only want normal termination".

So, hopefully this helps you understand why I have a problem with the status quo. My solution is hardly the only option, but I think it's pretty good.
Based on your criticism, here's an idea to improve it actually:

Would it be easier to understand if we made sure to wrap errors from resource clean-up vs archetype execution, so they say `error during cleanup of MyArchetype(%self): %w`, or `error during execution of ...`? Then, even if the errors come from the same control flow location, it's clear when looking at the error value what happened (e.g you get one list of clearly labeled error info, that you can, let's be honest, mostly print + panic, or deconstruct with `errors.Is` if you really want). The whole "return the errors as-is" concept comes from an early prototype of PGo, and there's no need to keep that unchanged... in fact, it's probably a source of confusion, tbh.

I hope I have somehow convinced you we should at least _change something_, if not my exact proposal.

Reply:
I see your point. First. I clarify that I didn't mean that `MPCalContext` and `context.Context` are the same. I just mention that `context.Context` as an example that allows you to distinguish what kind of errors happened in your function. If the function didn't do its normal execution, you can easily find out if it was because of context cancellation or the function itself got an error.

Your point with FFI is true. `MPCalContext` is not something reusable and it's better to do the cleanup after its execution, rather than forcing the user to manually do that. Although, I think `MPCalContext` is not exactly an FFI since it allocates some resources in the beginning (which is not only memory, it can be sockets for example) and it might get some errors during the cleanup. Since the user is explicitly involved in creating the resources, it also makes sense that he/she might need to see resources' cleanup errors apart from archetype execution error.

I see two possible solutions:
1. Making execution and cleanup errors transparent to user. New API does this, but the problem is that sometimes you want handle some errors differently (such as `ErrContextClosed`) and this API doesn't easily allow you to do that. So we have a new run function that makes the API inconsistent. I was thinking do we really want this `ErrContextClosed`? Because with current API user have to explicitly call `RequestExit` to trigger this error and in this case he/she easily can find out that the MPCal context stopped or not. When I added this error to the project, I was inspired by `context.Context` design. For `context.Context` makes sense to have this error, since it has some API that allows user to cancel the API in the future (such as `context.WithTimeout`). 
2. Make the different types of errors are completely visible to user by returning two separate errors. Maybe adding a new struct named `MPCalContextRunErrors` and include the errors into that.

I think solution 1 is very similar to the your current design and solution 2 is similar to your proposed solution. I'm not sure which one is better. First solution has a simpler error handling but the second has more explicit semantics. What do you think?

Reply:
> I didn't mean that MPCalContext and context.Context are the same

I understand, I was just trying to guess at some clue as to why we couldn't understand each other before. Based on the rest of your reply, glad something I guessed seemed to work.

> I think MPCalContext is not exactly an FFI

Sure, at least not a "normal" one.

> I see two possible solutions:

I think I'm leaning toward (1), partly because it's mostly already done, but also because it seems optimised for the common case (most of the time we just want a simple error case to catch everything).
I'm fine removing `ErrContextClosed`, at least by default. Maybe we could have a config option to re-enable it, because otherwise you really can't tell (due to concurrency) if the archetype terminated or was killed, and you might occasionally care (expecting non-kill outcome only, maybe useful in a couple of our tests).

So, to mostly repeat what you said, we have only `Run()`, so there's no ambiguity what to call. By default, it's easier to just return `nil` if the archetype is killed, like you suggest, as most invocations I think would be async "daemon-style".
My only change would be: if you're "expert" enough to know about the option, you can specify on context creation that you want to enable `ErrContextClosed`, which might be used once or twice in current codebase.

There's probably also some ways to improve the massive `multierr` that could be returned (indicating more what came from where), but maybe we can look at that later, as an improvement on the basic idea above.

Does that sound good?

Reply:
Sound good. I just suggest completely removing `ErrContextClosed` until we find a use case that really need it, then we can add the option. It's better to keep the API as simple as possible when there is no need for complexity.

Reply:
I couldn't find any use case for `ErrContextClosed`, so I removed it as you suggested.